Add end position

rjbou opened the pull request:
Changes types of OpamParserTypes, to `foo_kind` the main type value and `foo` the type  with a position

```
type file_name = string
type pos =
  { filename: file_name;
    start: int * int; (* line, column *)
    stop: int * int; (* line, column *)
  }
type 'a with_pos =
  { pelem : 'a;
    pos : pos
  }
type value_kind =
  | Int of int
  [...]
and value = value_kind with_pos
```

rjbou committed:
Add end position

* Changes types of OpamParserTypes, to `foo_kind` the main type value and `foo` the type  with a position

type file_name = string
type pos =
  { filename: file_name;
    start: int * int; (* line, column *)
    stop: int * int; (* line, column *)
  }
type 'a with_pos =
  { pkind : 'a;
    pos : pos
  }
type value_kind =
  | Int of int
  [...]
and value = value_kind with_pos

dra27 commented:
Sorry to only just be looking at this, but isn’t a problem to break anything which uses the library? It should be possible to have the parser produced the more-detailed tree as a new entry point and have a function to transform it using the name of the old entry point?

rjbou commented:
On the lib dependencies, I already ported most of [them](https://opam.ocaml.org/packages/opam-file-format/) (opam, opam-ed, conex, kept travis-opam), and a test suite is needed to ensure that nothing is broken (I discovered bugs with conex test suite :roll_eyes: ). 
We can't keep the same datatype, as position was given only on their starting char, and not for everything (value itself has no position, only the beginning one of its first component, etc.). But it is possible to add a compatibility module, to give some translation functions.

rjbou committed:
Add `FullPos` modules, that implements the new with position types

rjbou committed:
add deprecation notice for main functions

rjbou commented:
Updated. Some duplicated code to avoid double transformation.
Added also deprecation notice, maybe better to add it in another PR and keep it for another release, to let people the time to update before they have the deprecation error.

rjbou committed:
Port tests

rjbou commented:
rebased, test ported, merging once CI finish

dra27 commented on src/opamBaseParser.mly:
```diff
@@ -11,15 +11,27 @@
+      filename = spos.pos_fname;
+      start = spos.pos_lnum, spos.pos_cnum - spos.pos_bol;
+      stop = epos.pos_lnum, epos.pos_cnum - epos.pos_bol;
+    })
```
```suggestion
  let get_pos n = get_pos_full ~s:n n
```

dra27 commented on src/opamBaseParser.mly:
```diff
@@ -64,28 +78,46 @@ items:
+| atom relop atom            {{ pos = get_pos_full 3 ; pelem = Relop ($2,$1,$3) }}
+| atom envop atom            {{ pos = get_pos_full 3 ; pelem = Env_binding ($1,$2,$3) }}
+| pfxop value                {{ pos = get_pos_full 2 ; pelem = Pfxop ($1,$2) }}
+| relop atom                 {{ pos = get_pos_full 2 ; pelem = Prefix_relop ($1,$2) }}
```
```suggestion
| atom RELOP atom            {{ pos = get_pos_full 3 ; pelem = Relop ({ pos = get_pos 2 ; pelem = $2 },$1,$3) }}
| atom ENVOP atom            {{ pos = get_pos_full 3 ; pelem = Env_binding ($1,{ pos = get_pos 2 ; pelem = $2 },$3) }}
| PFXOP value                {{ pos = get_pos_full 2 ; pelem = Pfxop ({ pos = get_pos 1 ; pelem = $1 },$2) }}
| RELOP atom                 {{ pos = get_pos_full 2 ; pelem = Prefix_relop ({ pos = get_pos 1 ; pelem = $1 },$2) }}
```
...

dra27 commented on src/opamBaseParser.mly:
```diff
@@ -94,12 +126,21 @@ values:
+| STRING                     {{ pos = get_pos 1 ; pelem = String $1 }}
 ;
 
+relop:
+| RELOP                      {{ pos = get_pos 1 ; pelem = $1 }}
+
+envop:
+| ENVOP                      {{ pos = get_pos 1 ; pelem = $1 }}
+
+pfxop:
+| PFXOP                      {{ pos = get_pos 1 ; pelem = $1 }}
+
```
```suggestion
```
These cause shift/reduce conflicts - I think the code duplication can only be avoided with a Menhir-based parser. Someone will grumble...

rjbou replied:
I don't remember why I splitted them... ok to reintegrate them

dra27 replied:
For `relop` it feels like code duplication because it's done twice - but it's a limitation of `ocamlyacc` grammars that you can't effectively substitute the action back in

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -10,61 +10,124 @@
-(** Logical operators *)
-type logop = [ `And (** [&] *) | `Or (** [|] *) ]
+  (** Logical operators *)
+  type logop = [ `And (** [&] *) | `Or (** [|] *) ]
 
-(** Prefix operators *)
-type pfxop = [ `Not (** [!] *) | `Defined (** [?] *) ]
+  (** Prefix operators *)
+  type pfxop = [ `Not (** [!] *) | `Defined (** [?] *) ]
 
-type file_name = string
+  type file_name = string
 
-(** Source file positions: [(filename, line, column)] *)
-type pos = file_name * int * int
+  (** Source file positions: [(filename, line, column)] *)
+  type pos = file_name * int * int
 
-(** Environment variable update operators *)
-type env_update_op = Eq       (** [=] *)
-                   | PlusEq   (** [+=] *)
-                   | EqPlus   (** [=+] *)
-                   | ColonEq  (** [:=] *)
-                   | EqColon  (** [=:] *)
-                   | EqPlusEq (** [=+=] *)
+  (** Environment variable update operators *)
+  type env_update_op = Eq       (** [=] *)
+                     | PlusEq   (** [+=] *)
+                     | EqPlus   (** [=+] *)
+                     | ColonEq  (** [:=] *)
+                     | EqColon  (** [=:] *)
+                     | EqPlusEq (** [=+=] *)
+end
```
I'd revert this in its entirety (i.e. have these types exactly as before)

rjbou replied:
This module has been introduced to avoid duplication (& non equality) of types, and to avoid having ugly types aliasing, e.g. for `env_update_op`, I had 
```
  type _env_update_op = env_update_op
  type env_update_op_kind = _env_update_op
  and env_update_op = env_update_op_kind with_pos
```

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -10,61 +10,124 @@
+      pos : pos
+    }
+
+  type relop_kind = Common.relop
+  and rel_op = relop_kind with_pos
```
It's slightly strange having OpamParserTypes.relop and OpamParserTypes.FullPos.rel_op - it feels that it would be better that `open OpamParserTypes` and `open OpamParserTypes.FullPos` give as similar a signature as possible? What's wrong with:
```suggestion
  type relop_kind = relop
  type relop = relop_kind with_pos
```

rjbou replied:
In fact, this renaming don't originate from the `FullPos` modules [introduction](https://github.com/ocaml/opam-file-format/pull/24/commits/39e8100b401e0f8e0c9a826b6b05d5d38a3ecf3b#diff-034c9d7984c00f23420753a0c3c0f77402e6a87f5d068a8bb6103b80fa9a1798R27). It is an opam problem, as an include done in [`OpamTypes`](https://github.com/ocaml/opam/pull/4298/files#diff-011f82839082bb0a4d19ca919973f41e80fe5a7c37b6c849ebdd80325e8d73aeR24) and the type relop used as is for `formula`s & co.
And after that, It's just a cosmetic consistency change for `log_op` and `pfx_op`

dra27 replied:
Can it be fixed in `OpamTypes` instead?

rjbou replied:
If it is fixed in opam, we need to change the name of `relop` type. When I made those changes, I had in mind minimal changes in opam's api, only the external value type changed.

rjbou replied:
Or not include `OpamParserTypes`, and  create all needed aliases.

dra27 replied:
Indeed - just keeping all the alterations here as small as possible!

rjbou replied:
Ok, added some `open` in opamp, to remove the include, and then no more naming problem!

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -10,61 +10,124 @@
+  type relop_kind = Common.relop
+  and rel_op = relop_kind with_pos
+
+  type logop_kind = Common.logop
+  and log_op = logop_kind with_pos
```
Similarly, 
```suggestion
  type logop_kind = logop
  type logop = logop_kind with_pos
```

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -10,61 +10,124 @@
+  type logop_kind = Common.logop
+  and log_op = logop_kind with_pos
+
+  type pfxop_kind = Common.pfxop
+  and pfx_op = pfxop_kind with_pos
```
and:
```suggestion
  type pfxop_kind = pfxop
  type pfxop = pfxop_kind with_pos
```

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -10,61 +10,124 @@
+  type pfxop_kind = Common.pfxop
+  and pfx_op = pfxop_kind with_pos
+
+  type env_update_op_kind = Common.env_update_op
+  and env_update_op = env_update_op_kind with_pos
```
This doesn't work as intended without listing the constructors - I presume it's meant to be that we have `OpamParserTypes.FullPos.Eq`, etc.?). This creates a type alias, not an equivalent type, so you can have:

```ocaml
let x : OpamParserTypes.FullPos.env_update_op_kind = OpamParserTypes.Eq
(* or *)
let x : OpamParserTypes.FullPos.env_update_op_kind = OpamParserTypes.Common.Eq
```
but not
```ocaml
let x : OpamParserTypes.FullPos.env_update_op_kind = OpamParserTypes.FullPos.Eq
```
so:
```suggestion
  type env_update_op_kind = env_update_op =
    Eq | PlusEq | EqPlus | ColonEq | EqColon | EqPlusEq
  type env_update_op = env_update_op_kind with_pos
```

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -10,61 +10,124 @@
+    | Bool of bool
+    | Int of int
+    | String of string
+    | Relop of rel_op * value * value
+    | Prefix_relop of rel_op * value
+    | Logop of log_op * value * value
+    | Pfxop of pfx_op * value
```
```suggestion
    | Relop of relop * value * value
    | Prefix_relop of relop * value
    | Logop of logop * value * value
    | Pfxop of pfxop * value
```

dra27 commented on src/opamPrinter.mli:
```diff
@@ -16,34 +16,45 @@
 val opamfile: opamfile -> string
-(** Converts an {!opamfile} to a string. *)
+[@@ocaml.deprecated "Use OpamPrinter.FullPos.opamfile instead."]
+(** Converts an {!opamfile} to a string, using the
+    {!OpamParserTypes.opamfile.file_crlf} field to determine how to encode line
+    endings. *)
```
```suggestion
(** Converts an {!opamfile} to a string. *)
```
(the extra bit has crept in from #11 somehow!)

dra27 commented on src/opamPrinter.mli:
```diff
@@ -69,11 +80,17 @@ val format_opamfile: Format.formatter -> opamfile -> unit
 *)
 module Normalise : sig
   val escape_string : string -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise.escape_string instead."]
   val value : value -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise.value instead."]
   val item : opamfile_item -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise.item instead."]
   val item_order : opamfile_item -> opamfile_item -> int
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise.item_order instead."]
   val items : opamfile_item list -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise.items instead."]
   val opamfile : opamfile -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise.opamfile instead."]
 end
```
```suggestion
module Normalise : sig
  val escape_string : string -> string
  val value : value -> string
  val item : opamfile_item -> string
  val item_order : opamfile_item -> opamfile_item -> int
  val items : opamfile_item list -> string
  val opamfile : opamfile -> string
end
  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Normalise instead."]
```
The module deprecation has the benefit that it also displays on `open OpamPrinter.Normalise`

dra27 commented on src/opamPrinter.mli:
```diff
@@ -84,8 +101,10 @@ module Preserved : sig
       elements. The function assumes that [str] parses to the items
       [orig_its]. *)
   val items: string -> opamfile_item list -> opamfile_item list -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Preserved.items instead."]
 
   val opamfile: ?format_from:file_name -> opamfile -> string
+  [@@ocaml.deprecated "Use OpamPrinter.FullPos.Preserved.opamfile instead."]
   (** [opamfile f] converts [f] to string, respecting the layout and comments in
       the corresponding on-disk file for unmodified items. [format_from] can be
       specified instead of using the filename specified in [f]. *)
```
GitHub won't let me make the full suggestion here - but the same as with `Normalise`, just deprecate the entire module.

dra27 commented on src/opamPrinter.mli:
```diff
@@ -94,7 +113,53 @@ end
+
+  open OpamParserTypes.FullPos
+
+  val relop: [< relop_kind ] with_pos -> string
+  val logop: [< logop_kind ] with_pos -> string
+  val pfxop: [< pfxop_kind ] with_pos -> string
+  val env_update_op: env_update_op -> string
+
+  val relop_kind: [< relop_kind ] -> string
+  val logop_kind: [< logop_kind ] -> string
+  val pfxop_kind: [< pfxop_kind ] -> string
+  val env_update_op_kind: env_update_op_kind -> string
+
+  val value : value -> string
+  val value_list: value list with_pos -> string
+  val items: opamfile_item list -> string
+  val opamfile: opamfile -> string
+
+  val format_opamfile: Format.formatter -> opamfile -> unit
+
+  module Normalise : sig
+    val escape_string : string -> string
+    val value : value -> string
+    val item : opamfile_item -> string
+    val item_order : opamfile_item -> opamfile_item -> int
+    val items : opamfile_item list -> string
+    val opamfile : opamfile -> string
+  end
+
+  module Preserved : sig
+    val items: string -> opamfile_item list -> opamfile_item list -> string
+
+    val opamfile: ?format_from:file_name -> opamfile -> string
+  end
+
+  val value_equals: value -> value -> bool
+  val opamfile_item_equals: opamfile_item -> opamfile_item -> bool
```
It seems mean (!!) to have the comments on the deprecated functions above - perhaps just duplicate them to here? Equally for documentation, it might be worth putting this module first and then all the deprecated functions below it with a single comment about their oldness!

rjbou replied:
I went into minimal changes, with a full definition for the old definition, and `FullPos` only as a temporary change. But yes, it's better to have the comment in the new one.

dra27 commented on src/dune:
```diff
@@ -5,5 +5,8 @@
   (wrapped false)
   (modules_without_implementation opamParserTypes))
 
+(env (dev
+  (flags (:standard -warn-error -A))))
+
```
Does this really want to be committed?

rjbou replied:
it can be reverted...

dra27 commented on src/opamPrinter.ml:
```diff
@@ -324,3 +324,340 @@ module Preserved = struct
     items txt orig.file_contents f.file_contents
```
GitHub won't let me note this on the correct line, but the line above wants to be `let[@ocaml.warning "-3"] =`

dra27 commented on tests/oldpos.ml:
```diff
@@ -0,0 +1,490 @@
+(*  exception on linking described in the file LICENSE.                   *)
+(*                                                                        *)
+(**************************************************************************)
+
```
```suggestion

[@@@ocaml.warning "-3"]

```

dra27 reviewed (commented):
I'm sorry for taking so long to look at this again! The review looks horrific, but in fact it's mainly suggestions to eliminate the shift/reduce warnings in the parser, structure the types slightly differently in `OpamParserTypes` (I think it is an error that you can't have `OpamParserTypes.FullPos.Eq`, for example) and shift some doc comments around.

dra27 commented on tests/dune:
```diff
@@ -2,4 +2,5 @@
   (package opam-file-format)
-  (libraries alcotest opam-file-format))
+  (libraries alcotest opam-file-format)
+  (flags (:standard -warn-error -3)))
```
I think it's better to do this in `tests/oldpos.ml` so that it's just that one script which has it?

rjbou replied:
```suggestion
```

rjbou commented:
Thanks for the review!

rjbou committed:
Update comment & move them on full pos function instead of legacy

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -11,76 +11,148 @@
+module FullPos : sig
+
+  (** Source file positions *)
+  type file_name = Common.file_name
```
```suggestion
  type nonrec file_name = file_name
```

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -11,76 +11,148 @@
-                   | EqPlusEq (** [=+=] *)
-
-(** Base values *)
+(** Type definitions used by the legacy and the new full position modules *)
+module Common: sig
+
+  (** Relational operators *)
+  type relop = [ `Eq  (** [=] *)
+               | `Neq (** [!=] *)
+               | `Geq (** [>=] *)
+               | `Gt  (** [>] *)
+               | `Leq (** [<=] *)
+               | `Lt  (** [<] *)
+               ]
+
+  (** Logical operators *)
+  type logop = [ `And (** [&] *) | `Or (** [|] *) ]
+
+  (** Prefix operators *)
+  type pfxop = [ `Not (** [!] *) | `Defined (** [?] *) ]
+
+  type file_name = string
+
+  (** Source file positions: [(filename, line, column)] *)
+  type pos = file_name * int * int
+
+  (** Environment variable update operators *)
+  type env_update_op = Eq       (** [=] *)
+                     | PlusEq   (** [+=] *)
+                     | EqPlus   (** [=+] *)
+                     | ColonEq  (** [:=] *)
+                     | EqColon  (** [=:] *)
+                     | EqPlusEq (** [=+=] *)
+end
```
Putting these in a module isn't necessary - it just requires a `nonrec` or two.

```suggestion
(** Relational operators *)
type relop = [ `Eq  (** [=] *)
             | `Neq (** [!=] *)
             | `Geq (** [>=] *)
             | `Gt  (** [>] *)
             | `Leq (** [<=] *)
             | `Lt  (** [<] *)
             ]
 (** Logical operators *)
type logop = [ `And (** [&] *) | `Or (** [|] *) ]
 (** Prefix operators *)
type pfxop = [ `Not (** [!] *) | `Defined (** [?] *) ]
 type file_name = string
 (** Source file positions: [(filename, line, column)] *)
type pos = file_name * int * int
 (** Environment variable update operators *)
type env_update_op = Eq       (** [=] *)
                   | PlusEq   (** [+=] *)
                   | EqPlus   (** [=+] *)
                   | ColonEq  (** [:=] *)
                   | EqColon  (** [=:] *)
                   | EqPlusEq (** [=+=] *)
```

dra27 commented on src/opamParserTypes.mli:
```diff
@@ -11,76 +11,148 @@
+
+end
+
+include module type of struct include Common end
+
```
```suggestion
```

dra27 commented:
The build failure on 4.02 is a bit tedious, but can be fixed with:

```diff
diff --git a/src/dune b/src/dune
index 1c55e14..508e8d1 100644
--- a/src/dune
+++ b/src/dune
@@ -3,7 +3,16 @@
   (public_name opam-file-format)
   (synopsis "Parser and printer for the opam file syntax")
   (wrapped false)
+  (flags :standard (:include flags.sexp))
   (modules_without_implementation opamParserTypes))

+(rule
+  (with-stdout-to flags.ml
+    (echo "print_string (if String.sub Sys.ocaml_version 0 5 = \"4.02.\" then \"(-w -50)\" else \"()\")")))
+
+(rule
+  (with-stdout-to flags.sexp
+    (run ocaml %{dep:flags.ml})))
+
```
(unfortunately attributes are not processed during parsing)

dra27 commented on tests/fullpos.ml:
```diff
@@ -0,0 +1,524 @@
+
+  let print_value v =
+    OpamPrinter.FullPos.value v
+    |> String.split_on_char '\n'
```
```suggestion
    |> split_on_char '\n'
```

dra27 reviewed (commented):
Typo in tests/fullpos.ml is blocking the 4.02/4.03 tests passing.

Just need to eliminate `OpamParserTypes.Common` and rename the `rel_op`, etc. types and I think we're there

rjbou committed:
Rollback to old type names

AltGr commented on src/opamParserTypes.mli:
```diff
@@ -39,48 +41,114 @@ type env_update_op = Eq       (** [=] *)
+module FullPos : sig
+
+  (** Source file positions *)
+  type nonrec file_name = file_name
```
Requires OCaml >= 4.02.2

dra27 commented:
We're already in 4.02 territory with the .mli files because we're using attributes for the deprecations. Do we really need to worry about < 4.02.3?

rjbou commented:
In the test, it is 4.02.3 the oldest one

dra27 commented:
FWIW, this restores < 4.02.2 support 🙂 
```diff
diff --git a/src/Makefile b/src/Makefile
index e1eb5f6..d3c8c17 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -3,17 +3,31 @@ all: opam-file-format.cmxa opam-file-format.cma
 %.ml: %.mll
        ocamllex $<

+OCAML_VERSION:=$(shell ocamlc -vnum | tr -d '\r' | sed -e 's/+.*//' -e 's/\.//g')
+ifeq ($(OCAML_VERSION),)
+OCAML_VERSION:=0
+endif
+
+ifeq ($(shell test $(OCAML_VERSION) -ge 4022 || echo no-attributes),)
+PP=
+else
+PP=-pp 'sed -e s/\\[@@\\?ocaml[^]]*\\]//'
+endif
+
+show:
+       echo $(OCAML_VERSION)
+
 %.ml %.mli: %.mly
        ocamlyacc $<

 %.cmi: %.mli
-       ocamlc -c $<
+       ocamlc -c $(PP) $<

 %.cmo: %.ml %.cmi
-       ocamlc -c $<
+       ocamlc -c $(PP) $<

 %.cmx: %.ml %.cmi
-       ocamlopt -c $<
+       ocamlopt -c $(PP) $<

 MODULES = $(sort $(basename $(wildcard *.ml *.mll *.mly)))

@@ -41,6 +55,6 @@ clean:
        rm -f *.cm* *.o *.a $(GENERATED) .depend .merlin

 .depend: *.ml *.mli $(GENERATED)
-       ocamldep *.mli *.ml > .depend
+       ocamldep $(PP) *.mli *.ml > .depend

 -include .depend
diff --git a/src/opamParserTypes.mli b/src/opamParserTypes.mli
index 601201b..f57f3f3 100644
--- a/src/opamParserTypes.mli
+++ b/src/opamParserTypes.mli
@@ -45,7 +45,7 @@ type env_update_op = Eq       (** [=] *)
 module FullPos : sig

   (** Source file positions *)
-  type nonrec file_name = file_name
+  type file_name = string

   (** Full position *)
   type pos = {
```

rjbou committed:
lt 4.02.2 compatibility

rjbou committed:
4.02 compatibility

rjbou commented:
all green \o/
